event/datadeal/data_verification_event.go

```diff
@@ -1 +1,231 @@
+
+	dealID, err := strconv.ParseUint(dealIDStr, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	voteOption, err := d.verifyAndGetVoteOption(dealID, dataHash)
+	if err != nil {
+		if voteOption == oracletypes.VOTE_OPTION_UNSPECIFIED {
+			return fmt.Errorf("can't vote due to error while verify. dealID(%d). dataHash(%s)", dealID, dataHash)
+		} else {
+			log.Infof("vote NO due to error while verifying. dealID(%d). dataHash(%s)", dealID, dataHash)
+		}
+	}
+
+	msgVoteDataVerification, err := makeDataVerificationVote(
+		d.reactor.OracleAcc().GetAddress(),
+		dataHash,
+		dealID,
+		voteOption,
+		d.reactor.OraclePrivKey().Serialize(),
+	)
+	if err != nil {
+		return err
+	}
+
+	txBuilder := panacea.NewTxBuilder(*d.reactor.QueryClient())
+
+	txBytes, err := generateTxBytes(msgVoteDataVerification, d.reactor.OracleAcc().GetPrivKey(), d.reactor.Config(), txBuilder)
+	if err != nil {
+		return fmt.Errorf("generate tx failed. dealID(%d). dataHash(%s): %v", dealID, dataHash, err)
+	}
+
+	if err := broadcastTx(d.reactor.GRPCClient(), txBytes); err != nil {
+		return fmt.Errorf("broadcast transaction failed. dealID(%d). dataHash(%s): %v", dealID, dataHash, err)
+	}
+
+	return nil
+}
+
+func (d DataVerificationEvent) decryptData(decryptedSharedKey, nonce, encryptedDataBz []byte) ([]byte, error) {
+	decryptedData, err := crypto.DecryptWithAES256(decryptedSharedKey, nonce, encryptedDataBz)
+	if err != nil {
+		return nil, err
+	}
+	return decryptedData, nil
+}
+
+func (d DataVerificationEvent) compareDataHash(dataSale *types.DataSale, decryptedData []byte) bool {
+	decryptedDataHash := sha256.Sum256(decryptedData)
+	decryptedDataHashStr := hex.EncodeToString(decryptedDataHash[:])
+
+	return decryptedDataHashStr == dataSale.DataHash
+}
+
+func (d DataVerificationEvent) convertSellerData(deal *types.Deal, dataSale *types.DataSale) ([]byte, error) {
+	encryptedDataBz, err := d.reactor.Ipfs().Get(dataSale.VerifiableCid)
+	if err != nil {
+		log.Infof("failed to get data from IPFS: %v", err)
+		return nil, nil
+	}
+
+	oraclePrivKey := d.reactor.OraclePrivKey()
+
+	sellerAcc, err := d.reactor.QueryClient().GetAccount(dataSale.SellerAddress)
+	if err != nil {
+		log.Infof("failed to get account from panacea: %v", err)
+		return nil, nil
+	}
+
+	sellerPubKeyBytes := sellerAcc.GetPubKey().Bytes()
+
+	sellerPubKey, err := btcec.ParsePubKey(sellerPubKeyBytes, btcec.S256())
+	if err != nil {
+		log.Infof("failed to parsing sellerPubKey: %v", err)
+		return nil, nil
+	}
+
+	decryptSharedKey := crypto.DeriveSharedKey(oraclePrivKey, sellerPubKey, crypto.KDFSHA256)
+
+	decryptedData, err := d.decryptData(decryptSharedKey, deal.Nonce, encryptedDataBz)
+	if err != nil {
+		return nil, err
+	}
+
+	return decryptedData, nil
+}
+
+func (d DataVerificationEvent) verifyAndGetVoteOption(dealID uint64, dataHash string) (oracletypes.VoteOption, error) {
+	deal, err := d.reactor.QueryClient().GetDeal(dealID)
+	if err != nil {
```

Comment:
We decided to vote `VOTE_OPTION_NO` for a `NotFound` error during `GetDeal` or `GetDataSale` query. This is because it is difficult to see this situation as `oracle`'s fault.

Reply:
Sorry. Voting for `dataSale` that does not exist is meaningless, so it seems to be okay.